### PR TITLE
BUG: date_range linspace behavior respects tz

### LINKED ISSUE #20983: BUG: linspace behavior for date_range is incorrect when tz aware
#### Code Sample, a copy-pastable example if possible

xref #20846

```python
In [2]: dti = pd.date_range('20180101', '20180103', periods=3, tz='US/Eastern')

In [3]: dti
Out[3]: DatetimeIndex(['2017-12-31', '2018-01-01', '2018-01-02'], dtype='datetime64[ns, US/Eastern]', freq=None)

In [4]: dti[0]
Out[4]: Timestamp('2017-12-31 19:00:00-0500', tz='US/Eastern')
```
#### Problem description
The dates in the `date_range` are not being localized correctly; the dates are being shifted back by the UTC offset, then having the timezone appended to the shifted back date.

#### Expected Output
I'd expect this to be identical to the equivalent non-linspace behavior call:
```python
In [5]: dti2 = pd.date_range('20180101', periods=3, freq='D', tz='US/Eastern')

In [6]: dti2
Out[6]:
DatetimeIndex(['2018-01-01 00:00:00-05:00', '2018-01-02 00:00:00-05:00',
               '2018-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [7]: dti2[0]
Out[7]: Timestamp('2018-01-01 00:00:00-0500', tz='US/Eastern', freq='D')
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: b3f07b2551b7eb40408ca28416ca5e21f8fb37b4
python: 3.6.1.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.0rc2+23.gb3f07b2
pytest: 3.1.2
pip: 9.0.1
setuptools: 39.0.1
Cython: 0.28.2
numpy: 1.13.3
scipy: 1.0.0
pyarrow: 0.6.0
xarray: 0.9.6
IPython: 6.1.0
sphinx: 1.5.6
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: 0.4.0
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 0.9.8
lxml: 3.8.0
bs4: None
html5lib: 0.999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
fastparquet: 0.1.5
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #20983
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry (addendum to the v0.23 whatsnew entry for this feature)

### CODE REVIEW

**Comment by jschendel** [comment]
Beat me by a few minutes!

Can you modify the test that was originally written for this issue:

https://github.com/pandas-dev/pandas/blob/b3f07b2551b7eb40408ca28416ca5e21f8fb37b4/pandas/tests/indexes/datetimes/test_date_range.py#L165-L180

The first half looks to be redundant with your new test.  The timezone transition portion isn't being tested, as there's no `tm.assert_index_equal`.  It also looks wrong to me; periods is set to three, but the expected has 5 elements.  I think the correct expected is:

```python
DatetimeIndex(['2018-04-01 01:00:00+11:00', '2018-04-01 02:00:00+10:00',
               '2018-04-01 04:00:00+10:00'],
              dtype='datetime64[ns, Australia/Sydney]', freq=None)
```

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 1]
```diff
@@ -179,6 +179,18 @@ def test_date_range_convenience_periods(self):
...
 
+    @pytest.mark.parametrize('start,end', [
+        ['20180101', '20180103'],
+        [datetime(2018, 1, 1), datetime(2018, 1, 3)],
+        [Timestamp('20180101'), Timestamp('20180103')],
+        [Timestamp('20180101', tz='US/Eastern'),
+         Timestamp('20180103', tz='US/Eastern')]])
+    def test_date_range_linspacing_tz(self, start, end):
+        # GH 20983
+        result = date_range(start, end, periods=3, tz='US/Eastern')
```
Comment (jschendel) [inline pandas/tests/indexes/datetimes/test_date_range.py]: Might be beneficial to test when `start`/`end` are tz aware but not specifying the `tz` param for `date_range`, to ensure that `tz` inference from `start`/`end` is okay.

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -587,10 +587,13 @@ def _generate(cls, start, end, periods, name, freq,
                     if end is not None:
                         end = end.tz_localize(tz).asm8
             else:
+                # Create a linearly spaced date_range.
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: specifiy this is in local time

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small comment by @jschendel

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke and @jschendel

**Pull request merged**